networkx/algorithms/isomorphism/tests/VF2++/test_vf2pp.py

```diff
@@ -0,0 +1,411 @@
+import networkx as nx
+import random
+from networkx.algorithms.isomorphism.VF2pp import isomorphic_VF2pp
+
+
+def assign_labels(G1, G2, mapped_nodes=None):
+    colors = [
+        "white",
+        "black",
+        "green",
+        "purple",
+        "orange",
+        "red",
+        "blue",
+        "pink",
+        "yellow",
+        "none",
+    ]
+    for node in G1.nodes():
+        color = colors[random.randrange(0, len(colors))]
+        G1.nodes[node]["label"] = color
+        if mapped_nodes:
+            node = mapped_nodes[node]
+        G2.nodes[node]["label"] = color
+
+    return G1, G2
+
+
+def get_labes(G1, G2):
+    return nx.get_node_attributes(G1, "label"), nx.get_node_attributes(G2, "label")
+
+
+class TestVF2pp:
+    def test_both_graphs_empty(self):
+        G = nx.Graph()
+        H = nx.Graph()
+        isomorphic, mapping = isomorphic_VF2pp(G, H, {}, {})
+        assert isomorphic
+        assert mapping == {}
+
+    def test_first_graph_empty(self):
+        G = nx.Graph()
+        H = nx.Graph([(0, 1)])
+        isomorphic, mapping = isomorphic_VF2pp(G, H, {}, {})
+        assert not isomorphic
+        assert mapping is None
+
+    def test_second_graph_empty(self):
+        G = nx.Graph([(0, 1)])
+        H = nx.Graph()
+        isomorphic, mapping = isomorphic_VF2pp(G, H, {}, {})
+        assert not isomorphic
+        assert mapping is None
+
+    def test_disconnected_graph(self):
+        num_nodes = [100, 330, 579, 631, 799]
+        for Vi in num_nodes:
+            nodes = [i for i in range(Vi)]
+            G1 = nx.Graph()
+            G2 = nx.Graph()
+
+            G1.add_nodes_from(nodes)
+            G2.add_nodes_from(nodes)
+
+            G1, G2 = assign_labels(G1, G2)
+            l1, l2 = get_labes(G1, G2)
+
+            isomorphic, mapping = isomorphic_VF2pp(G1, G2, l1, l2)
+            assert isomorphic
+            assert len(set(mapping)) == G1.number_of_nodes()
+
+    def test_custom_graph1(self):
+        G1 = nx.Graph()
+        G2 = nx.Graph()
+
+        edges1 = [(1, 2), (1, 3), (1, 4), (2, 3), (3, 4), (5, 1), (5, 2), (5, 4)]
+        edges2 = [
+            ("A", "B"),
+            ("A", "C"),
+            ("A", "D"),
+            ("B", "C"),
+            ("C", "D"),
+            ("Z", "A"),
+            ("Z", "B"),
+            ("Z", "D"),
+        ]
+        G1.add_edges_from(edges1)
+        G2.add_edges_from(edges2)
+
+        mapped = {1: "A", 2: "B", 3: "C", 4: "D", 5: "Z"}
+        G1, G2 = assign_labels(G1, G2, mapped)
+        l1, l2 = get_labes(G1, G2)
+
+        isomorphic, mapping = isomorphic_VF2pp(G1, G2, l1, l2)
+        assert isomorphic
+        assert mapping == mapped
```

Comment:
Is it possible to guarantee this order? `{1: "A", 2: "B", 3: "C", 4: "D", 5: "Z"}`
`1` should be mapped to `A` but other nodes can be mapped to anything else right?
That seems to be the test failure in https://github.com/networkx/networkx/runs/7370107803?check_suite_focus=true

Reply:
If order of a dict is causing trouble you can test each element like:
```python
assert all(mapping[n] == mapped[n] for n in mapped) and len(mapping) == len(mapped)
```

Reply:
oops... I misunderstood.  This isn't a problem of the dict having the wrong order. The problem is that there can be more than one mapping that works. 

When a network has symmetries, there will be multiple solutions. So, to get a test with a unique solution, we should add or remove edges until the symmetries are removed.  Here, we could leave out edge (4, 5) and add edge (2, 6).  

To make the test more obviously checking that an isomorphism __does__ exist, we could set up this example by creating G1 and then using `G2 = nx.relabel_nodes(G1, mapped)`.  Then since we set them up to be isomorphic, the test is more clear. It's the same test -- just written differently.

Reply:
@MridulS @dschult You are both right. I thought that in the graph that i created there's only one possible mapping but i am wrong (and locally it never fails so i didn't notice). I will update the test right now.

Reply:
Also, do you think it's better to remove the exhaustive unit tests and only keep the small ones?

Reply:
Small tests that are easy to understand are often better than large tests where it isn't clear what they are testing. Of course, it is good to have both while developing the code. But it might be good to remove the tests that we can't describe what features they are testing.